packages/dds/tree/src/feature-libraries/persisted-config/persistedConfig.ts

```diff
@@ -0,0 +1,336 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+
+// TODO: Document, decide if semver-like even provides value here (over e.g. plain-old number).
+export type Version = `${number}.${number}.${number}${"-internal" | ""}`;
+
+export interface PersistedConfig {
+	version: Version;
+	// Strictly increasing number which identifies the iteration of the proposed protocol. It's suggested
+	// that applications are able to modify this using the same mechanisms with which they can modify a DDS's
+	// configuration, which enables safe rollback policies.
+	// Example: application enables feature flag 1 in iteration 2, then realizes there is a problem with it.
+	// application then disables feature flag 1 in iteration 3.
+	// Using a single version number and monotonic increase on it isn't sufficient: DDSes may support configuration
+	// which impacts their serialization format with independent semantics (ex: attribution, policies to include history)
+	protocolIteration: number;
+	flags: Record<string, boolean>; // TODO: consider making this generic
+}
+
+export enum ConfigUpgradeType {
+	ConcurrentOpsInvalid,
+	ConcurrentOpsValid,
+}
+
+// note: would be great to lock this down. changing this for flags that have been added can have nasty side effects
+// for in progress upgrades.
+// TODO: document process for adding flags
+export interface PersistedConfigSchema {
+	version: (current: Version, previous: Version) => ConfigUpgradeType;
```

Comment:
How does the application declare what configurations it supports? I think an important feature of this library is having nice errors when an app encounters (due to loading a document or a config change whiles its connected) a change which it doesn't understand. I'd expect the App to implement this and return some error (or undefined) here to indicate such a case, but the signature does not allow that.

Reply:
> I think an important feature of this library is having nice errors when an app encounters (due to loading a document or a config change whiles its connected) a change which it doesn't understand.

absolutely agree this is a goal. I think I was originally intending this error to be surfaced in `onProtocolChange`, but it's an interesting thought to do so in response to this function's return value: we could certainly generate an informative error. I'll give it a try.

Worth noting that this mechanism isn't intended to be used by apps directly. In its current shape, it would be the DDS that specifies something like maximum versions that it supports.